starknet_py/net/full_node_client.py

```diff
@@ -0,0 +1,267 @@
+            )
+
+        if block_hash is None and block_number is None:
+            raise ValueError("Block_hash or block_number must be provided.")
+
+        res = None
+        if block_hash is not None:
+            res = await self.rpc_client.call(
+                method_name="getBlockByHash",
+                params={
+                    "block_hash": str(hex(block_hash))
+                    if isinstance(block_hash, int)
+                    else block_hash,
+                    "requested_scope": "FULL_TXNS",
+                },
+            )
+        if block_number is not None:
+            res = await self.rpc_client.call(
+                method_name="getBlockByNumber",
+                params={"block_number": block_number, "requested_scope": "FULL_TXNS"},
+            )
+        return StarknetBlockSchema().load(res, unknown=EXCLUDE)
+
+    async def get_state_update(
+        self,
+        block_hash: Optional[Union[int, str]] = None,
+        block_number: Optional[int] = None,
+    ) -> BlockState:
+        # TODO when pathfinder node adds support (currently untestable)
+        pass
+
+    async def get_storage_at(
+        self,
+        contract_address: Union[int, str],
+        key: int,
+        block_hash: Optional[Union[int, str]] = None,
+        block_number: Optional[int] = None,
+    ) -> int:
+        if block_hash is None:
+            raise ValueError("Block_hash must be provided when using FullNodeClient.")
+
+        res = await self.rpc_client.call(
+            method_name="getStorageAt",
+            params={
+                "contract_address": str(hex(contract_address))
+                if isinstance(contract_address, int)
+                else contract_address,
+                "key": str(hex(key)),
+                "block_hash": str(hex(block_hash))
+                if isinstance(block_hash, int)
+                else block_hash,
+            },
+        )
+        res = typing.cast(str, res)
+        return int(res, 16)
+
+    async def get_transaction(
+        self,
+        tx_identifier: Union[
+            Union[int, str], BlockHashIdentifier, BlockNumberIdentifier
+        ],
+    ) -> Transaction:
+        res = None
+        error_message = None
+
+        if isinstance(tx_identifier, (int, str)):
+            res = await self._get_transaction_by_tx_hash(tx_identifier)
+
+        try:
+            block_hash_identifier = BlockHashIdentifierSchema().load(tx_identifier)
+            res = await self._get_transaction_by_block_hash(block_hash_identifier)
+        except ValidationError as ex:
+            error_message = error_message or str(ex)
+
+        try:
+            block_number_identifier = BlockNumberIdentifierSchema().load(tx_identifier)
+            res = await self._get_transaction_by_block_number(block_number_identifier)
+        except ValidationError as ex:
+            error_message = error_message or str(ex)
```

Comment:
This doesn't look good.

Reply:
I agree it's not the most elegant way to write but here's a problem: user passes a block_hash_identifier or block_number_identifier, which is a typeddict. So for python it's just a dict so to check what private method to call, we either have to use these schemas or manually check the keys in provided dict.

Or we can change this parameters entirely (how?)

Reply:
That makes sense. I think I focused too much on making the interface as inclusive as possible and these kind of problems could be resolved with just simplifying the interface. I'll try this approach for this and maybe some other methods.